[TECH] Marquer le bon déploiement Scalingo comme nécessaire au merge

VincentHardouin opened the pull request:
## :unicorn: Problème
Le bon déploiement des RA n'est pas bloquant pour le merge, ce qui est problématique par exemple pour l'auto-merge des PRs renovate ou encore lors d'un rebase fait pas l'action `auto-merge`. De plus, il est difficile de marquer facilement le déploiement de la RA sur Scalingo comme bloquant, car le nom du check étant différent pour chaque PR nous pouvons pas l'ajouter aux checks obligatoire dans les protections de branches. 

![Screenshot 2024-09-24 at 12 15 16](https://github.com/user-attachments/assets/8d3eb6a8-c4db-44f4-bc29-aa68cb28e291)

De plus, le déploiement de la RA sur Scalingo, ne fait pas partie de la check-suite, mais fait partie des status du commits (voir [cette réponse StackOverflow](https://stackoverflow.com/questions/67919168/github-checks-api-vs-check-runs-vs-check-suites/72312617#72312617) pour comprendre la différence), ce qui rend la tâche plus compliqué. 

## :robot: Proposition


## :rainbow: Remarques
<!-- Des infos supplémentaires, trucs et astuces ? -->

## :100: Pour tester
<!-- Les instructions pour reproduire le problème, les profils de test, le parcours spécifique à utiliser, etc. -->

pix-service-auto-merge commented:
:tada: This issue has been resolved in version 0.4.0 :tada:

The release is available on [GitHub release](https://github.com/1024pix/pix-api-data/releases/tag/v0.4.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

VincentHardouin committed:
feat: wait for status checks

GitHub allows waiting for check suites, but does not allow waiting for commit status checks,
such as the successful deployment of Scalingo.

pix-bot-github commented:
Une fois l'application déployée, elle sera accessible à cette adresse https://api-data-pr37.review.pix.fr
Les variables d'environnement seront accessibles sur scalingo https://dashboard.scalingo.com/apps/osc-fr1/pix-api-data-integration-pr37/environment

Gudsfile commented on .github/workflows/wait-for-status-checks.yml:
```diff
@@ -0,0 +1,42 @@
+          echo "Listen for commit $COMMIT"
+          WAITED="0"
+          SLEEP_TIME="60"
+          while [ "$STATUS" == "pending" ] && [ "$WAITED" -le 240 ]
```
Pourquoi forcément attendre 240sec, c'est pour debug ?

VincentHardouin replied:
Nope, c'est vraiment pour attendre, l'action se lance dès le push du dernier commit, mais il faut bien attendre que tous les status soient finis, le déploiement met environ 2-3 min sur Scalingo, donc il faut attendre autant de temps : 
![Screenshot 2024-09-23 at 14 46 46](https://github.com/user-attachments/assets/98f3a182-b6a3-4963-b88f-51a53a111594)

Gudsfile replied:
Ah oui j'ai mal lu, donc la ci attend un statut autre que "pending" pendant max 240sec. Si au bout des 240sec le status est toujours pending ça fail.

Cette durée est propre au projet, il faudrait peut-être en faire une variable de manière à ce qu'elle soit plus visible, qu'en penses-tu ?

Gudsfile replied:
Quid des cas où le déploiement est très long (exemple lorsqu'il est en fil d'attente), il faudra relancer cette tâche avant de pouvoir merge c'est ça ?

VincentHardouin replied:
Yes exactement !

Gudsfile commented on .github/workflows/wait-for-status-checks.yml:
```diff
@@ -0,0 +1,42 @@
+
+      - name: Use correct sha
+        run: |
+          echo "before: $GITHUB_SHA"
+          if [ "$GITHUB_EVENT_NAME" == "pull_request" ]; then
+           GITHUB_SHA=${{ github.event.pull_request.head.sha }}
+          fi
+          echo COMMIT_SHA=$GITHUB_SHA >> $GITHUB_ENV
+          echo "after: $GITHUB_SHA"
```
Tu sais pourquoi ce n'est pas le bon sha au départ ?

VincentHardouin replied:
[Github dans la doc ](https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request) dit que la valeur de `GITHUB_SHA` quand c'est un évènement de type `pull_request` équivaut au sha du commit de merge, seulement ce n'est pas le commit de merge qui a les status mais bien le dernier commit de la branche. 

![Screenshot 2024-09-23 at 14 45 09](https://github.com/user-attachments/assets/5b9e9baf-2b13-4b48-8005-cf0de86cfd33)

GradedJestRisk commented on .github/workflows/wait-for-status-checks.yml:
```diff
@@ -0,0 +1,41 @@
+          do
+            sleep ${SLEEP_TIME}
+            WAITED=$((WAITED+SLEEP_TIME))
+            STATUS=$(gh api /repos/${{ github.repository  }}/commits/"${COMMIT_SHA}"/status -q .state)
```
:exploding_head: Inception !!

VincentHardouin commented:
Ce n'est pas cette manière qui me semble la plus pérenne. Ici, nous devons connaitre le temps de build du projet, et implémenter ça dans chaque repo. De plus cela fait énormement consommé des minutes de GitHub Actions inutilement.

Je travaillerai sur le fait que Pix Bot au moment de déployer une RA créé un check et le change son status une fois la RA déployée.
